coalib/misc/CachingUtilities.py

```diff
@@ -0,0 +1,208 @@
+
+
+def time_consistent(log_printer, project_hash):
+    """
+    Verify if time is consistent with the last time was run. That is,
+    verify that the last run time is in the past. Otherwise, the
+    system time was changed and we need to flush the cache and rebuild.
+
+    :param log_printer:  A LogPrinter object to use for logging.
+    :param project_hash: A MD5 hash of the project directory to be used
+                         as the key.
+    :return:             Returns True if the time is consistent and as
+                         expected; False otherwise.
+    """
+    time_db = pickle_load(log_printer, "time_db", {})
+    if project_hash not in time_db:
+        # This is the first time coala is run on this project, so the cache
+        # will be new automatically.
+        return True
+    return time_db[project_hash] <= calendar.timegm(time.gmtime())
+
+
+def update_time_db(log_printer, project_hash, current_time=None):
+    """
+    Update the last run time on the project.
+
+    :param log_printer:  A LogPrinter object to use for logging.
+    :param project_hash: A MD5 hash of the project directory to be used
+                         as the key.
+    :param current_time: Current time in epoch format. Not giving this
+                         argument would imply using the current system time.
+    """
+    if not current_time:
+        current_time = calendar.timegm(time.gmtime())
+    time_db = pickle_load(log_printer, "time_db", {})
+    time_db[project_hash] = current_time
+    pickle_dump(log_printer, "time_db", time_db)
+
+
+def get_last_mod_time(files):
+    """
+    Determine the last modified time of the given ``files``.
+
+    :param files: A list of file paths.
+    :return:      Return the modification time of the last modified
+                  file.
+    """
+    mod_time = -1
+    for config in files:
+        if os.path.isfile(config):
+            # The last changed file among all files should suffice since
+            # it would trigger the same consequences as any specific file.
+            mod_time = max(mod_time, int(os.path.getmtime(config)))
+    return mod_time
+
+
+def config_file_changed(log_printer, project_hash, project_dir):
+    """
+    Determine if the config file is new or has changed since the last run.
+
+    :param log_printer:  A LogPrinter object to use for logging.
+    :param project_hash: A MD5 hash of the project directory to be used
+                         as the key.
+    :param project_dir:  The root directory of the project to be used
+                         as a key identifier.
+    :return:             Return True if the config file has changed
+                         Return False otherwise.
+    """
+    # When the -c option is given, project_dir points to the given config file.
+    # So we must check project_dir and project_dir + default_coafile.
+    config_files = [os.path.join(project_dir, default_coafile),
+                    project_dir,
```

Comment:
CC @sils1297 

This should solve the `-c` config issue we were discussing, I think.